File: bumperbot_python_examples/bumperbot_python_examples/simple_qos_publisher.py
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSDurabilityPolicy
from std_msgs.msg import String #Library for Message type which will be published

class SimpleQoSPublisher(Node):
    def __init__(self):
        super().__init__("simple_qos_publisher")

        self.qos_profile_pub = QoSProfile(depth=10) #Make a quality of service profile

        self.declare_parameter("reliability", "system_default") #Here reliability parameter is set to system default
        self.declare_parameter("durability", "system_default") #Here durability parameter is set to system default

        reliability = self.get_parameter("reliability").get_parameter_value().string_value
        durability = self.get_parameter("durability").get_parameter_value().string_value

        if reliability == "best_effort":
            self.qos_profile_pub.reliability = QoSReliabilityPolicy.BEST_EFFORT
            self.get_logger().info(f"[Reliability] : {reliability}")
        elif reliability == "reliable":
            self.qos_profile_pub.reliability = QoSReliabilityPolicy.RELIABLE
            self.get_logger().info(f"[Reliability] : {reliability}")
        elif reliability == "system_default":
            self.qos_profile_pub.reliability = QoSReliabilityPolicy.SYSTEM_DEFAULT
            self.get_logger().info(f"[Reliability] : {reliability}")
        else:
            self.get_logger().error(f"Selected Reliability QoS \"{reliability}\" doesn't exist")
            return
        
        if durability == "volatile":
            self.qos_profile_pub.durability = QoSDurabilityPolicy.VOLATILE
            self.get_logger().info(f"[Durability] : {durability}")
        elif durability == "transient_local":
            self.qos_profile_pub.durability = QoSDurabilityPolicy.TRANSIENT_LOCAL
            self.get_logger().info(f"[Durability] : {durability}")
        elif durability == "system_default":
            self.qos_profile_pub.durability = QoSDurabilityPolicy.SYSTEM_DEFAULT
            self.get_logger().info(f"[Durability] : {durability}")
        else:
            self.get_logger().error(f"Selected Durability QoS \"{durability}\" doesn't exist")
            return

        


        self.pub_ = self.create_publisher(String, "chatter", self.qos_profile_pub)

        self.counter_ = 0
        self.frequency_ = 1.0

        self.get_logger().info(f"Publishing at {self.frequency_} Hz")
        self.timer_ = self.create_timer(self.frequency_, self.timerCallback) #function which sets the frequency of the publisher or subscriber messages

    def timerCallback(self):
        msg = String()
        msg.data = f"Hello ROS2 - counter: {self.counter_}" #this sets up the string message data to be published

        self.pub_.publish(msg) #this call will publish the message
        self.counter_ += 1

    
def main():
    rclpy.init()
    simple_qos_publisher = SimpleQoSPublisher()
    rclpy.spin(simple_qos_publisher) #this makes the thread continuously run the publisher class object
    simple_qos_publisher.destroy_node() #this will destroy the publisher node smoothly when Ctrl+C pressed
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: bumperbot_cpp_examples/src/simple_qos_publisher.cpp
#include <rclcpp/rclcpp.hpp> //Include the ros C++ library
#include <std_msgs/msg/string.hpp> //Include the standard message type string

#include <chrono>

using namespace std::chrono_literals;

class SimpleQoSPublisher : public rclcpp::Node
{
public:
    SimpleQoSPublisher() : Node("simple_qos_publisher"), qos_profile_pub_(10), counter_(0)
    {
        declare_parameter<std::string>("reliability", "system_default"); //initiation of parameter delcaration
        declare_parameter<std::string>("durability", "system_default");

        const auto reliability = get_parameter("reliability").as_string(); //variable setting of parameter
        const auto durability = get_parameter("durability").as_string();

        //setting up reliability as per defined
        if(reliability == "best_effort"){
            qos_profile_pub_.reliability(RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT);
            RCLCPP_INFO(get_logger(), "[Reliability] : %s", reliability.c_str());
        }
        else if(reliability == "reliable"){
            qos_profile_pub_.reliability(RMW_QOS_POLICY_RELIABILITY_RELIABLE);
            RCLCPP_INFO(get_logger(), "[Reliability] : %s", reliability.c_str());
        }
        else if(reliability == "system_default"){
            qos_profile_pub_.reliability(RMW_QOS_POLICY_RELIABILITY_SYSTEM_DEFAULT);
            RCLCPP_INFO(get_logger(), "[Reliability] : %s", reliability.c_str());
        }
        else {
            RCLCPP_ERROR_STREAM(get_logger(), "Selected Reliability QoS: " << reliability << " doesn't exists!");
        }

        //setting up durability as per defined
        if(durability == "volatile"){
            qos_profile_pub_.durability(RMW_QOS_POLICY_DURABILITY_VOLATILE);
            RCLCPP_INFO(get_logger(), "[Reliability] : %s", durability.c_str());
        }
        else if(durability == "transient_local"){
            qos_profile_pub_.durability(RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL);
            RCLCPP_INFO(get_logger(), "[Reliability] : %s", durability.c_str());
        }
        else if(durability== "system_default"){
            qos_profile_pub_.durability(RMW_QOS_POLICY_DURABILITY_SYSTEM_DEFAULT);
            RCLCPP_INFO(get_logger(), "[Reliability] : %s", durability.c_str());
        }
        else {
            RCLCPP_ERROR_STREAM(get_logger(), "Selected Reliability QoS: " << durability << " doesn't exists!");
        }


        pub_ = create_publisher<std_msgs::msg::String>("chatter", qos_profile_pub_);
        timer_ = create_wall_timer(1s, std::bind(&SimpleQoSPublisher::timerCallback, this));

        RCLCPP_INFO(get_logger(), "publishing at 1 Hz");
    }

private:
    unsigned int counter_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::QoS qos_profile_pub_;

    void timerCallback()
    {
        auto message = std_msgs::msg::String();
        message.data = "Hello ROS 2 - counter: " + std::to_string(counter_++);

        pub_->publish(message);
    }
};



int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleQoSPublisher>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}